### DISCUSSION TITLE
RE: performance

### DISCUSSION THREAD

**Discussion by Joe23232** [gfx-rs/wgpu#2080] [General]
Hi, in regards to the performance of wgpu:

Is it slower than say vulkin? If so is it due to the fact that it is an abstraction layer or is it due to the fact that it is all written in Safe Rust which has its limits or is it due to both reasons?

**Comment by kvark** [comment]
Hi!

In theory, it is slower than raw Vulkan for the following reasons:
  1. it lays over Vulkan, so it can't be faster that raw Vulkan
  2. it doesn't expose all of the platform-specific features, so one could do things with raw Vulkan that they can't do with wgpu today (e.g. ray tracing)

Nothing about Safe Rust makes it slower, fwiw.

Now, in practice, our abstraction over Vulkan is very efficient. You can see it in `wgpu-hal`, which is a layer of abstraction that is portable but unsafe. If you check out the repo and do `cargo run --release --example halmark`, then hit space a few times to spawn many sprites, you'll see when it starts being laggy. I don't think you can get faster with raw Vulkan (by doing the same thing).
However, wgpu also adds validation, state tracking, and lifetime tracking. These have a cost, and we are constantly working on reducing it. The worst case for this cost is that it makes your CPU usage twice as slow, as indicated by comparing `halmark` example with `bunnymark` (which uses full wgpu). In a real app, I'd expect the overhead be 5-10%.

**Reply by Joe23232** [reply]
I see thanks mate :)

**Reply by Joe23232** [reply]
Hey mate  @kvark 

> However, wgpu also adds validation, state tracking, and lifetime tracking.

Can this be disabled when using WGPU?

**Reply by kvark** [reply]
No, this is integral to wgpu-core. The state tracking is not only needed for validation, it's also to generate proper Vulkan barriers and DX12 resource transitions. The lifetime tracking is to know when to delete objects.

**Reply by Joe23232** [reply]
I see thanks mate

**Reply by jorgeperezlara** [reply]
I understand `wgpu-hal` doesn't have validation, state tracking, or lifetime tracking, you may use that, @Joe23232

**Comment by Joe23232** [comment]
I see thanks mate
